Add Timeouts Blogpost (#479)
temporalio/documentation: blog/2021-06-17-activity-timeouts.md

  values={[
    {label: 'Go', value: 'go'},
    {label: 'Java', value: 'java'},
  ]
}>

<TabItem value="go">

```go
var result string
err := workflow.ExecuteActivity(ctx, SimpleActivity, value).Get(ctx, &result)
if err != nil {
        return err
}
```

</TabItem>
<TabItem value="java">

```java
@WorkflowInterface
public interface SimpleWorkflow {
    @WorkflowMethod
    String simpleWorkflowMethod(String someArg);
}

public static class SimpleWorkflowImpl implements SimpleWorkflow {
    private final SimpleActivities activities =
        Workflow.newActivityStub(
            SimpleActivities.class,
            ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(2)).build());

    @Override
    public String simpleWorkflowMethod(String name) {
      return activities.simpleActivity("Hello", name);
    }
}
```

</TabItem>
</Tabs>

Behind the scenes, the SDK transforms this to a `ScheduleActivity` Command, which is sent to the Temporal Server. This Command includes various metadata, including the activity type (`Foo`), activity task queue (`Bar`), activity ID, and `RetryPolicy` (if not specified, Temporal uses a default policy).

### Step 2 - Temporal Server

> Receiving the Command, Temporal Server then sets up the mutable state for that workflow and activity ID, and also adds an `ActivityTask` to the `Bar` Activity Queue.

Author: as dominik mentioned - better to just drop mutable state mention here altogether as not directly relevant to timeouts

  ]
}>

<TabItem value="go">

```go
progress := 0
for hasWork {
    // Send heartbeat message to the server.
    activity.RecordHeartbeat(ctx, progress)
    // Do some work.
    ...
    progress++
}
```

</TabItem>
<TabItem value="java">

```java
while ((read = inputStream.read(bytes)) != -1) {
  totalRead += read;
  f.write(bytes, 0, read);
  // Let the Server know about the download progress.
  Activity.getExecutionContext().heartbeat(totalRead);
}
```

</TabItem>
</Tabs>

Setting a `HeartbeatTimeout` allows us to retry activities more quickly (e.g. the next minute after a heartbeat is missed) than the `StartToCloseTimeout`, which is necessarily set to as long as the longest possible activity (e.g. five hours later when we are sure the activity should have been completed).

There are some minor nuances to heartbeats that may be of interest:

- You can freely record heartbeats as often as you want - once a minute, or everytime a loop iterates - the SDKs throttle the heartbeats that get sent back anyway.
- If a `HeartbeatTimeout` isn't set and the activity tries to record one, nothing will be recorded since that information will never be used.

## ScheduleToStart Timeout

> We use the `ScheduleToStartTimeout`  to set a maximum limit that an activity should sit in a queue. We recommend that most users monitor the `ScheduleToStart` latency metric and set alerts for them as a [production scaling](https://docs.temporal.io/docs/server/production-deployment/#faq-autoscaling-workers-based-on-task-queue-load) metric, rather than setting an explicit timeout for it.

Reviewer: ```suggestion
The Schedule-To-Start Timeout  sets a limit on the amount of time that an Activity Task can sit in a Task Queue.
We recommend that most users monitor the `ScheduleToStart` latency metric and set alerts for that as a [production scaling](https://docs.temporal.io/docs/server/production-deployment/#faq-autoscaling-workers-based-on-task-queue-load) metric, rather than setting an explicit timeout for it.
```
> `ScheduleToStart` latency metric

How is that actually emitted? `schedule-to-start` all lower case I am guessing?

---
tags:
  - Temporal
  - architecture
  - timeouts
posted_on_: 2021-06-17T00:00:09Z
slug: activity-timeouts

> title: 'The 4 Types of Activity Timeouts'

Reviewer 2: @flossypurse: Title style in docs is sentence case. Blog-post titles are currently a mix of sentence case and title case. Going forward, should we use sentence case for the blog as well as the docs?

Side note: Usually, style calls for spelling out numbers zero through nine. But for blog-post title, I'm inclined to stet.

Reviewer: ```suggestion
title: 'The 4 Types of Activity timeouts'
```

author: swyx
author_title: Head of Developer Experience
author_image_url: https://avatars.githubusercontent.com/u/6764957?v=4
release_version: V1.10.2
---

<!--truncate-->

> Part of the benefit of moving business logic to Temporal is how it implements retries and timeouts for activities in a standardized way. This has the effect of adding a reliability layer atop unreliable activities and workers, in a durable and scalable fashion. However, understanding the terminology can be a bit intimidating at first glance. 

Reviewer 2: ```suggestion
Part of the benefit of moving business logic to Temporal is how it implements retries and timeouts for Activities in a standardized way. This has the effect of adding a reliability layer atop unreliable Activities and Workers, in a durable and scalable fashion. However, understanding the terminology can be a bit intimidating at first glance. 
```

> This post (together with the embedded talk) aims to give you a solid mental model on what each activity timeout does and when to use it.

Reviewer 2: ```suggestion
This post (together with the embedded talk) aims to give you a solid mental model on what each Activity timeout does and when to use it.
```

> > Note: There are also [workflow timeouts](https://docs.temporal.io/docs/concepts/workflows#timeout-settings) and [retry policies](https://docs.temporal.io/docs/concepts/activities#retries) you can set. This post only deals with *activity* timeouts.

Reviewer 2: ```suggestion
> Note: You can also set [Workflow timeouts](https://docs.temporal.io/docs/concepts/workflows#timeout-settings) and [retry policies](https://docs.temporal.io/docs/concepts/activities#retries) you can set. This post deals only with *Activity* timeouts.
```

> ## Talk Version: Whiteboard Session

Reviewer: ```suggestion
## Talk version: whiteboard session
```

> You can watch this 18 minute talk where our CEO Maxim Fateev explains the 4 Types of Activity Timeouts you see in Temporal.

Reviewer 2: ```suggestion
In this 18-minute talk, our CEO Maxim Fateev explains the 4 types of Activity timeouts you see in Temporal.
```

import { ResponsivePlayer } from '../src/components'

<ResponsivePlayer url='https://www.youtube.com/watch?v=JK7WLK3ZSu8' />

The rest of this post summarizes the main points covered in this whiteboard session, for those who prefer a written version.

> ## Written Version: TL;DR

Reviewer: ```suggestion
## Written version: TL;DR
```

![image](https://user-images.githubusercontent.com/6764957/122424123-0ab1e300-cfc1-11eb-82c5-cf8856a87242.png)

> Activities go through 3 main states in Temporal: 

Reviewer 3: Activity Executions and Activity Task Executions go through their own respective state machines, but that warrants its own blog post

Author: ok I think i can just discard this para and we should follow up with a blogpost on these two

> Activities go through 3 main states in Temporal: 

Reviewer 2: ```suggestion
Activities go through three main states in Temporal: 
```

> Activities go through 3 main states in Temporal: 
> 
> - Scheduled
> - Started
> - Closed
> 

Author: drop the emphasis on the simplified state machine as we want to expand in future blogposts

```suggestion
```

> There are 4 Timeouts in Temporal — 2 that are commonly used, and 2 that are only useful in specific cases:

Reviewer 2: ```suggestion
Temporal has four timeouts—two that are commonly used, and two that are useful only in specific cases:
```

- `ScheduleToClose`: to limit maximum execution time including retries
- `StartToClose`: to limit maximum execution time of a single invocation. **We recommend ALWAYS setting this!**

> - `Heartbeat`: *For long running activities*, to respond faster when a regular heartbeat fails to be recorded.

Reviewer 2: ```suggestion
- `Heartbeat`: *For long-running activities*, to respond faster when a regular heartbeat fails to be recorded.
```

@flossypurse: Hyphenating "long-running" doesn't match our current usage, but I'm recommending the change. Stet if you want to hold off for now.

> - `ScheduleToStart`: *For queue timeouts and task routing*, to limit maximum time that an activity waits in a task queue. **This is rarely needed!**

Reviewer 2: ```suggestion
- `ScheduleToStart`: *For Task Queue timeouts and Task routing*, to limit maximum time that an Activity waits in a Task Queue. **This is rarely needed!**
```

> You can find the precise APIs in each SDK's reference documentation: [Java](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/activity/ActivityOptions.Builder.html) and [Go](https://pkg.go.dev/go.temporal.io/sdk@v1.7.0/internal#ActivityOptions).

Reviewer 2: ```suggestion
You can find the precise APIs in the reference documentation for each SDK: [Java](https://www.javadoc.io/doc/io.temporal/temporal-sdk/latest/io/temporal/activity/ActivityOptions.Builder.html) and [Go](https://pkg.go.dev/go.temporal.io/sdk@v1.7.0/internal#ActivityOptions).
```

## Lifecycle of an Activity

> To really understand how timeouts work, we should understand the typical lifecycle of an activity as it journeys through the various parts of the system.

Reviewer 2: ```suggestion
To really understand how timeouts work, we should understand the typical lifecycle of an Activity as it journeys through the various parts of the system.
```

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

### Step 1 - Workflow Worker

> An activity `Foo` is first invoked inside of a Workflow Worker on Task Queue `Bar`. The precise method of invocation differs by SDK, and timeouts are also specified up front as part of activity options:

Author: Foo and Bar are from Maxim's talk, i made up the code samples to flesh out to blog post. its easy to change:

```suggestion
An activity `SimpleActivity` is first invoked inside of a Workflow Worker on Task Queue `sampleTaskQUeue`. The precise method of invocation differs by SDK, and timeouts are also specified up front as part of activity options:
```

> An activity `Foo` is first invoked inside of a Workflow Worker on Task Queue `Bar`. The precise method of invocation differs by SDK, and timeouts are also specified up front as part of activity options:

Reviewer 2: Just FYI: Both [Google Dev Doc Style Guide](https://developers.google.com/style/word-list?hl=en#foo) and [Microsoft Writing Style Guide](https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/f/foo-foobar-fubar) recommend not using _foo_ and _bar_.

        // ScheduleToStartTimeout: time.Second * 60, // usually not needed! see below
        StartToCloseTimeout:    time.Second * 60,
        HeartbeatTimeout:       time.Second * 10,
        WaitForCancellation:    false,
}
ctx = workflow.WithActivityOptions(ctx, ao)

var result string
err := workflow.ExecuteActivity(ctx, SimpleActivity, value).Get(ctx, &result)
if err != nil {
        return err
}
```

</TabItem>
<TabItem value="java">

```java
@WorkflowInterface
public interface SimpleWorkflow {
    @WorkflowMethod
    String simpleWorkflowMethod(String someArg);
}

public static class SimpleWorkflowImpl implements SimpleWorkflow {
    private final SimpleActivities activities =
        Workflow.newActivityStub(
            SimpleActivities.class,
            ActivityOptions.newBuilder()
              .setScheduleToCloseTimeout(Duration.ofSeconds(500))
              .setStartToCloseTimeout(Duration.ofSeconds(60))
              .setHeartbeatTimeout(Duration.ofSeconds(10))
              .build());

    @Override
    public String simpleWorkflowMethod(String name) {
      return activities.simpleActivity("Hello", name);
    }
}
```

</TabItem>
</Tabs>

> Behind the scenes, the SDK transforms this to a `ScheduleActivity` Command, which is sent to the Temporal Server. This Command includes various metadata, including the activity type (`Foo`), activity task queue (`Bar`), activity ID, and `RetryPolicy` (if not specified, Temporal uses a default policy).

Author: ```suggestion
Behind the scenes, the SDK transforms this to a `ScheduleActivity` Command, which is sent to the Temporal Server. This Command includes various metadata, including the activity type (`SimpleActivity`), activity task queue (`sampleTaskQueue`), and activity ID. A `RetryPolicy` was not specified, so Temporal uses [the default Retry Policy](https://docs.temporal.io/docs/concepts/workflows#retry-policy).
```

### Step 3 - Activity Worker 

An Activity Worker that has been polling for the `Bar` activity queue picks up the `ActivityTask` and begins execution.

> The activity is now in `STARTED` state.

### Step 4 - Temporal Server

Once the activity finishes successfully, the Activity Worker sends a `CompleteActivityTask` message (together with the result of the activity) to Temporal Server, which now gives control back to the Workflow Worker to continue to the next line of code and repeat the process.

> The activity is now in `CLOSED` state.

We have just described the "Happy Path" of an activity. However, what happens when a worker crashes midway through an execution?

## StartToClose Timeout

We use the `StartToCloseTimeout` to control the maximum amount of time **a single activity invocation** can take. We recommend **always** setting this timeout.

![image](https://user-images.githubusercontent.com/6764957/122290108-fff64000-cf25-11eb-92b3-0533e41c3fee.png)

The classical example for how `StartToClose` becomes relevant is when an activity has been picked up from the activity queue (`STARTED` state) but crashes midway (so it never reaches `CLOSED` state). 

> - Without a timeout configured, Temporal would never proactively timeout this activity to initiate a retry. The activity becomes "stuck" and the end user experiences an indefinite delay of their work with no feedback.

Reviewer 3: An activity cannot be stuck if (ScheduleToClose timeout configured) or ((ScheduleToStart timeout and StartToClose timeout configured) or (ScheduleToStart timeout and Heartbeat timeout configured))

There are some minor nuances to heartbeats that may be of interest:

- You can freely record heartbeats as often as you want - once a minute, or everytime a loop iterates - the SDKs throttle the heartbeats that get sent back anyway.
- If a `HeartbeatTimeout` isn't set and the activity tries to record one, nothing will be recorded since that information will never be used.

## ScheduleToStart Timeout

We use the `ScheduleToStartTimeout`  to set a maximum limit that an activity should sit in a queue. We recommend that most users monitor the `ScheduleToStart` latency metric and set alerts for them as a [production scaling](https://docs.temporal.io/docs/server/production-deployment/#faq-autoscaling-workers-based-on-task-queue-load) metric, rather than setting an explicit timeout for it.

![image](https://user-images.githubusercontent.com/6764957/122290279-287e3a00-cf26-11eb-8dd6-3133016a0bd9.png)

> As a queue timeout, `ScheduleToStart` is unique in that it **cannot be retried** — all a retry would do is pop the activity right back on to the same queue!

Reviewer 3: ScheduleToClose will not be retried either

> As a queue timeout, `ScheduleToStart` is unique in that it **cannot be retried** — all a retry would do is pop the activity right back on to the same queue!

Reviewer 3: We have not introduced queue timeout and non-queue timeouts

> The `ScheduleToStartTimeout` is most useful when you have a concrete plan to reroute activity to a different queue, if a given task queue is not draining in adequate time. You can also reschedule a whole set of other activities or do other compensation logic based on this timeout. This is a powerful feature for building ultra-reliable systems, however most users will not need this since you can horizontally scale the number of workers easily.

Reviewer: ```suggestion
The Schedule-To-Start Timeout is most useful when you have a concrete plan to reroute an Activity Task to a different Task Queue, if a given Task Queue is not draining in adequate time.
You can also reschedule a whole set of other Activity Executions or do other compensation logic based on this timeout.
This is a powerful feature for building ultra-reliable systems, however most users will not need this since you can horizontally scale the number of Workers easily.
```
